### ISSUE TITLE
Reset namespace for testing config docs

### ISSUE THREAD

**Issue by bsipocz** [astropy/astropy#8781]
Currently doctesting has been switched off for `ScienceState` related docs, but ideally anything that suppose to work for the users should be tested. To make this happen we need to reset the namespace while testing, I'm certain there is a clever way to do it.

Bottom line, those docs should be tested by the time we ship it to the users, if nothing else by brute force removing the modules prior in a `testsetup` directive, etc.

 See discussion here: https://github.com/astropy/astropy/pull/8517#discussion_r289228267

**Comment by pllim** [comment]
`ScienceState` was originally added by @mdboom in #2094. As far as I can understand the implementation, unlike `ConfigItem`, it has not `reset` method. That is, once you call `set`, the given value is all it remembers. (Please correct me if I am wrong.)

Therefore, to enable what you propose, @stargaser might have to extend his state sub-classes to remember some kind of "original values" for the `reset` step.

Or maybe it was a concious design decision to not have a `reset`? I don't know.

**Comment by bsipocz** [comment]
No, what I mean here is on the pytest level. If you copy paste the examples for that given narrative docs, there are not issues whatsoever, but there are when running the full test suite. So for the sake of the users we need to test the functionality they normally encounter.

E.g. adding this  to the top of the file fixes it, but it's a bit too hacky, so I'm not given yet up to find a better solution.
```
.. testsetup::
   >>> import sys
   >>> del sys.modules['astropy.units']
   >>> del sys.modules['astropy.constants']
```

**Comment by stargaser** [comment]
> Could you comment on the failure I'm seeing in https://github.com/astropy/astropy/pull/8517#issuecomment-497876801?

_Originally posted by @bsipocz in https://github.com/astropy/astropy/pull/8517#issuecomment-502798267_

While working on #8517 I found that I could manipulate `sys.modules` to make the constant-version ScienceStates think that units and constants had not been imported. However this caused problems in pytests for other sub-packages like units. Most importantly, I could not get the check of `M_sun` to work with pytest. It seems one can fool `sys.modules` but those manipulations do not completely undo a prior import of units or constants.

The setting of constants versions works in a new Python or IPython session.

**Comment by bsipocz** [comment]
@stargaser - I strongly feel that these should be sorted out before 4.0 as the current way I don't see how well the behaviour is predictable in e.g. a pipeline, aka we should provide reproducible examples in the docs.

**Comment by pllim** [comment]
Maybe we should proceed with the `testsetup` solution as @bsipocz proposed in https://github.com/astropy/astropy/issues/8781#issuecomment-499207499 . We can figure out a better way to do this later.

**Comment by stargaser** [comment]
This might not help the doctest problem, but I think there is a way to write a better test that pytest can handle.

In file `check_msun.py`:
```
def check_msun_2014():
    import astropy
    astropy.physical_constants.set('codata2014')
    import astropy.constants as const
    import astropy.units as u
    assert abs((const.M_sun / u.M_sun).to(u.dimensionless_unscaled) - 1) < 1.e-14
    assert const.M_sun.reference[-11:] == 'CODATA 2014'
```

Run in a subprocess in file `test_setversion.py`:
```
import subprocess
import os


def test_versioning():
    path = os.path.join(os.path.dirname(__file__), "check_msun.py")
    status = subprocess.call(["python", path])
    assert status == 0
```

The `testsetup` solution of deleting units and constants from `sys.modules` will work for testing `physical_constants.set` but it doesn't work for the M_sun check that is [now in the units docs](https://astropy.readthedocs.io/en/latest/units/constants_versions.html). Maybe the combination of these two will get us close enough to ensuring the functionality is tested?

**Comment by stargaser** [comment]
Hmm, maybe a self-contained version for the docs can be made, writing the lines in the body of `check_msun_2014` to a temporary file, and then executing that temporary script in a subprocess. That would emphasize to users that changing the version only works in a new Python process (before any import of units or constants).

**Comment by pllim** [comment]
Re: separate script for doc -- It certainly has a precedence. `wcs` has example scripts in https://github.com/astropy/astropy/tree/master/docs/wcs/examples .

**Comment by pllim** [comment]
Taking liberty to move the milestone...

**Comment by bsipocz** [comment]
I haven't given up on it yet, but agree to remove the milestone.

**Comment by embray** [comment]
Assuming this still hasn't been fixed, I have some ideas for this issue.  I think we can take advantage of some more modern Python features (e.g. [PEP-562](https://www.python.org/dev/peps/pep-0562/)) to clean up some of this mess.

**Comment by pllim** [comment]
I don't think it has been fixed.
